File: src/output.rs
use std::path::{Path, PathBuf};

pub fn write_temp_html(html: &str) -> Result<PathBuf, String> {
    let tmpdir = std::env::var("TMPDIR").unwrap_or_else(|_| "/tmp".to_string());
    write_temp_html_in(html, &tmpdir)
}

pub fn write_temp_html_in(html: &str, tmpdir: &str) -> Result<PathBuf, String> {
    let timestamp = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map_err(|e| format!("clock error: {e}"))?
        .as_millis();
    let filename = format!("planview-{timestamp}.html");
    let path = Path::new(tmpdir).join(filename);
    std::fs::write(&path, html).map_err(|e| format!("cannot write '{}': {e}", path.display()))?;
    Ok(path)
}

pub fn open_browser(path: &Path) -> Result<(), String> {
    if cfg!(target_os = "windows") {
        std::process::Command::new("cmd")
            .args(["/C", "start", "", &path.display().to_string()])
            .spawn()
            .map_err(|e| format!("failed to open browser: {e}"))?;
    } else {
        let cmd = if cfg!(target_os = "macos") {
            "open"
        } else {
            "xdg-open"
        };
        std::process::Command::new(cmd)
            .arg(path)
            .spawn()
            .map_err(|e| format!("failed to open browser with '{cmd}': {e}"))?;
    }

    Ok(())
}

File: ts/output.ts
import { spawn } from "node:child_process";
import { writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

export function writeTempHtml(html: string, dir: string = tmpdir()): string {
  const timestamp = Date.now();
  const path = join(dir, `jidoka-${timestamp}.html`);
  writeFileSync(path, html);
  return path;
}

export function openBrowser(path: string): void {
  const [cmd, args] =
    process.platform === "win32"
      ? ["cmd", ["/C", "start", "", path]]
      : process.platform === "darwin"
        ? ["open", [path]]
        : ["xdg-open", [path]];
  // `spawn(...).unref()` lets us exit before the opener finishes, but a missing
  // binary fires an async `error` event — without this listener Node would
  // crash the parent with an unhandled exception even though our own work is
  // done. Swallow the failure so headless or stripped-down environments don't
  // fail an otherwise-successful render.
  const child = spawn(cmd as string, args as string[], {
    detached: true,
    stdio: "ignore",
  });
  child.on("error", (err) => {
    process.stderr.write(`jidoka: could not open browser (${cmd}): ${err.message}\n`);
  });
  child.unref();
}
